roachrest, roachprod: support no disks on workload machines

herkolategan opened the pull request:
Previously, workload machines automatically inherited the same disk setup as the rest of the cluster. To optimize for cost, we now allow workload machines to have only boot disks.

The default behavior, for providers, is to attach a standard data disk, or a number of local SSDs depending on the cluster spec. Thus in order to explicitly prevent any data disk from being configured , a new provider option "BootDiskOnly" has been added to prevent the default behavior, specifically for workload machines.

Epic: None
Release note: None

herkolategan committed:
roachprod: only wait on boot disk

Previously, during the wait for VMs to start we would check both the boot disk
and the data disk. But in order to support VMs that only have a boot disk we
should not be checking the data disk here. This change updates `Wait` to only
check the boot disk.

Epic: None
Release note: None

herkolategan committed:
roachtest: support boot disk only workload VMs

Previously, workload machines automatically inherited the same disk setup as the
rest of the cluster. To optimize for cost, we now allow workload machines to
attach only boot disks.

The default behavior is to attach a standard data disk. Hence, a new explicit
flag "BootDiskOnly" has been added to the provider options to prevent the
default behavior, specifically for workload machines.

Epic: None
Release note: None

herkolategan commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -3349,10 +3349,7 @@ type kafkaManager struct {
-	if k.c.IsLocal() {
-		return `/tmp/confluent`
-	}
-	return `/mnt/data1/confluent`
```
Hopefully, the OS volume size should be suffecient, if not we can bump that.

golgeek replied:
Maybe we could still create `/mnt/data1` in `setup_disk()` if `BootDiskOnly` is set?

golgeek replied:
Ah sorry, took a few minutes to think about and forgot to report back.

The main drawback I see with this is that this creates a potential for filling the boot disk if the developer assumes this is a different FS.

But fixing this would require creating another filesystem, which I guess would somewhat look like this:
```
mkdir -p /var/lib/workloads
fallocate -l 10G /var/lib/workloads/data1.img
mkfs.ext4 /var/lib/workloads/data1.img
mount -o loop /var/lib/workloads/data1.img /mnt/data1
```

Not sure it's really necessary...

herkolategan committed:
roachprod: add boot disk only cli flags

Epic: None
Release note: None

herkolategan committed:
roachtest: create data dir without disk

Previously, some tests will fail because they still expect `/mnt/data1` to be
present even if it is not used as a node disk, but for other reasons like
installing Kafka, or a directory based request run across all nodes including
the workload node.

Hence, we now create the directory even if the VM only has a boot disk to
support these scenarios.

Epic: None
Release note: None

herkolategan commented on pkg/cmd/roachtest/spec/cluster_spec.go:
```diff
@@ -286,7 +295,29 @@ func getGCEOpts(
+	}
+	opts.SSDCount = 0
+	opts.PDVolumeCount = 0
+	opts.BootDiskOnly = true
```
Ah right, I misunderstood that part.

It make sense adding `ClusterSpec.WorkloadRequiresDisk`. There are some tests which utilise a disk on a workload machine; I have added the option.

Having said that, I do think the usage of data disks on workload machines came from the fact that those also automatically got one, and in hind sight if a workload requires a bit more space it should actually just have a larger os volume.

But for now I'm happy with some cost savings. I also think there's quite a few tests that do not explicitly specify a workload node, even though it has one; so we still need to update those to start getting the full benefit.
